### PR TITLE
Skip adjustment item if originalInvoice reference id is missing

### PR DESCRIPTION
**Pull request by singh-sukhvir**
There are scenarios like below where adjustment item does not have linked OriginalInvoiceReferenceId -

1. Organization is not taxable and has recurring subscription from 15-15 every month.
2. It was billed on Jan 1 with target date as EOM.
3. Generated Invoice billed it from Dec15-Jan15 as expected with no taxes as its not taxable org.
4. Org A cancels subscription on 4th Jan.
5. Before Next billing i.e. Feb 1, it's decided that Org A must be taxed.
6. Billing on Feb-1 tries to repair to refund amount from 5th Jan to EOM. However Avatax tries to generate adjustment tax item for which Original tax item does not exist because Org was not taxable at that time.
7. This results in illegalStateException in Avatax plugin.

This PR not exactly fixes this but introduces a new plugin config property  "org.killbill.billing.plugin.avatax.adjustments.lenientMode" which if set to true would result in skipping such anomalous adjustment-Item.

### CODE REVIEW

File: src/main/java/org/killbill/billing/plugin/avatax/client/AvaTaxClient.java [hunk 1]
```diff
@@ -50,6 +50,8 @@
     private final String companyCode;
     private final String sanitizedCompanyCode;
     private final boolean commitDocuments;
+    // When true - skips adjustments which don't have corrrespnding previousInvoiceId
```
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/avatax/client/AvaTaxClient.java:53]: Typo: `corrrespnding`.

File: src/main/java/org/killbill/billing/plugin/avatax/api/AvaTaxTaxCalculator.java [hunk 2]
```diff
@@ -203,10 +207,22 @@ private CreateTransactionModel toTaxRequest(final String companyCode,
...
+                                                final LocalDate utcToday,
+                                                final boolean skipAnomalousAdjustments) {
+        
+        try {
+            Preconditions.checkState((originalInvoiceReferenceCode == null && (adjustmentItems == null || adjustmentItems.isEmpty())) ||
+                                        (originalInvoiceReferenceCode != null && (adjustmentItems != null && !adjustmentItems.isEmpty())),
+                                        "Invalid combination of originalInvoiceReferenceCode %s and adjustments %s", originalInvoiceReferenceCode, adjustmentItems);
+        } catch (IllegalStateException e) {                                     
+            if (skipAnomalousAdjustments) {
+                logger.warn("Returning null tax request instead of throwing IllegalStateException exception for Invalid combination of originalInvoiceReferenceCode {} and adjustments {}",
```
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/avatax/api/AvaTaxTaxCalculator.java:219]: The log message makes sense from a code perspective, but isn't really easy to understand when looking at it in tools like Splunk. What about something like:

```
logger.warn("Ignoring tax request due to inconsistent adjustments: originalInvoiceReferenceCode={}, adjustmentItems={}", originalInvoiceReferenceCode, adjustmentItems);
```

File: src/main/java/org/killbill/billing/plugin/avatax/client/AvaTaxClient.java [hunk 3]
```diff
@@ -50,6 +50,8 @@
...
+    private final boolean skipAnomalousAdjustments;
```
Comment (pierre) [inline src/main/java/org/killbill/billing/plugin/avatax/client/AvaTaxClient.java:54]: Could you also update the `README` documentation?

**Review by pierre — CHANGES_REQUESTED** [review]
Unfortunately, CI cannot run integration tests as secrets aren't propagated to PRs. Can you run them locally manually? You just need to create `src/test/resources/avatax.properties` with something like:
```
org.killbill.billing.plugin.avatax.url=https://sandbox-rest.avatax.com/api/v2
org.killbill.billing.plugin.avatax.accountId=XXX
org.killbill.billing.plugin.avatax.licenseKey=XXX
org.killbill.billing.plugin.avatax.companyCode=XXX
org.killbill.billing.plugin.avatax.commitDocuments=false
```
You can ignore the tests related to TaxRateApi.

File: src/main/java/org/killbill/billing/plugin/avatax/api/AvaTaxTaxCalculator.java (hunk 2 above)
Comment (singh-sukhvir) [reply src/main/java/org/killbill/billing/plugin/avatax/api/AvaTaxTaxCalculator.java:219]: this looks better, thanks for the suggestion. will update

**Comment by singh-sukhvir** [comment]
@ are the 

> Unfortunately, CI cannot run integration tests as secrets aren't propagated to PRs. Can you run them locally manually (`mvn test -Pintegration-mysql`)? You just need to create `src/test/resources/avatax.properties` with something like:
> 
> ```
> org.killbill.billing.plugin.avatax.url=https://sandbox-rest.avatax.com/api/v2
> org.killbill.billing.plugin.avatax.accountId=XXX
> org.killbill.billing.plugin.avatax.licenseKey=XXX
> org.killbill.billing.plugin.avatax.companyCode=XXX
> org.killbill.billing.plugin.avatax.commitDocuments=false
> ```
> 
> You can ignore the tests related to TaxRateApi.

Tests ran successfully locally 16/16 passed. But I have a question related to failures `ci/circleci: test-mysql ` && `ci/circleci: test-postgresql`. So my understanding based on your comment above is that, currently CI fails to run integration tests as secrets aren't propagated to PRs. Is that why these tests are failing in my PR?

File: src/main/java/org/killbill/billing/plugin/avatax/api/AvaTaxTaxCalculator.java [hunk 4]
```diff
@@ -203,10 +207,22 @@ private CreateTransactionModel toTaxRequest(final String companyCode,
...
                                                 final Iterable<PluginProperty> pluginProperties,
-                                                final LocalDate utcToday) {
-        Preconditions.checkState((originalInvoiceReferenceCode == null && (adjustmentItems == null || adjustmentItems.isEmpty())) ||
-                                 (originalInvoiceReferenceCode != null && (adjustmentItems != null && !adjustmentItems.isEmpty())),
-                                 "Invalid combination of originalInvoiceReferenceCode %s and adjustments %s", originalInvoiceReferenceCode, adjustmentItems);
+                                                final LocalDate utcToday,
+                                                final boolean skipAnomalousAdjustments) {
+        
+        try {
+            Preconditions.checkState((originalInvoiceReferenceCode == null && (adjustmentItems == null || adjustmentItems.isEmpty())) ||
```
Comment (sbrossie) [inline src/main/java/org/killbill/billing/plugin/avatax/api/AvaTaxTaxCalculator.java:214]: The original Precondition was a high level sanity check and I understand why in that scenario we would hit it, and therefore why introducing a  lenient mode (`skipAnomalousAdjustments`) avoids hitting such condition. But is this enough?

**Comment by pierre** [comment]
> But I have a question related to failures `ci/circleci: test-mysql ` && `ci/circleci: test-postgresql`. As you stated `currently CI fails to run integration tests as secrets aren't propagated to PRs.`, is that why these tests are failing in my PR?

Yes, you can see the 401 errors in the CircleCI logs:
image

I have now enabled GitHub actions on pull requests: https://github.com/killbill/killbill-avatax-plugin/commit/028e30e19131d4dfb3287c6c117588f356072b52

Could you merge master into your branch? With GitHub actions, one can *trust* a committer and propagate the secrets to PR runs. That way, we should be able to have a clean run (we are slowly moving away from CircleCI anyways).

**Comment by pierre** [comment]
Unfortunately, it turns out secrets aren't passed even if _trusted_ (https://securitylab.github.com/research/github-actions-preventing-pwn-requests/).

Let's ignore CI then for now since you ran them locally.

File: src/main/java/org/killbill/billing/plugin/avatax/client/AvaTaxClient.java (hunk 1 above)
Comment (singh-sukhvir) [reply src/main/java/org/killbill/billing/plugin/avatax/client/AvaTaxClient.java]: corrected in latest commit

File: src/main/java/org/killbill/billing/plugin/avatax/api/AvaTaxTaxCalculator.java (hunk 2 above)
Comment (singh-sukhvir) [reply src/main/java/org/killbill/billing/plugin/avatax/api/AvaTaxTaxCalculator.java]: updated in latest commit

File: src/main/java/org/killbill/billing/plugin/avatax/client/AvaTaxClient.java (hunk 3 above)
Comment (singh-sukhvir) [reply src/main/java/org/killbill/billing/plugin/avatax/client/AvaTaxClient.java:54]: updated in latest commit

**Review by pierre — APPROVED** [review]

**Comment by pierre** [comment]
avatax-plugin `0.8.5-8e71564` has been published.

**Comment by singh-sukhvir** [comment]
> avatax-plugin `0.8.5-8e71564` has been published.

awesome, thanks

**Pull request merged**